lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+				propagateIfChanged(after,res);
+			}
+
+			// At the end we can take the information about the arguments in our lattice and attach it
+			// to the function arguments
+			// TODO: understand when this is called and where to append the information obtained when finishing
+			// For the moment this function is just observing the IR, so we can leave it as it is
+			void setToExitState(ConstraintsLattice* lattice) override
+			{
+				
+				llvm::outs()<<"something was set to exit state\n";
+				
+			}
+
+			public:
+
+			void printRanges(mlir::Operation* op){
+				auto* lattice = op != nullptr ? getLattice(mlir::ProgramPoint(op))
+																			: getLattice(op->getBlock());
+				llvm::outs()<<"KEK ";
+				lattice->print(llvm::outs());
+			}
+
+			const llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange>& getUnderlyingLattice(mlir::Operation* op){
+				return this->getLattice(mlir::ProgramPoint(op))->getUnderlyingLattice();
+			}
+		};
+
+		// TODO: should I put it in another namespace ? (as in "ActionLiveness.cpp")
+		// This is the class that starts the analysis (I guess)
+		class Constraints
+		{
+			public:
+			explicit Constraints(mlir::rlc::FlatFunctionOp fun)
+					: function(fun), config()
+			{
+				config.setInterprocedural(false);
+				solver = std::make_unique<DataFlowSolver>(config);
+				//TODO: understand what are these -> other analysis? so should I keep them ? I think I should
+				solver->load<mlir::dataflow::DeadCodeAnalysis>();
+				solver->load<mlir::dataflow::SparseConstantPropagation>();
+				analysis = solver->load<ConstraintsAnalysis>(symbolTable);
+				auto top = fun;
+				llvm::outs()<<"Initialize analysis from operation: ";
+				top->print(llvm::outs());
+				auto res = solver->initializeAndRun(top);
+			}
+			
+			void printRangesFound(mlir::Operation* op){
+				llvm::outs()<<"Printing operation: ";
+				op->print(llvm::outs());
+				llvm::outs()<<"\n";
+				this->analysis->printRanges(op);
+			}
+
+			const llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange>& getOperationLattice(mlir::Operation* op){
+				return this->analysis->getUnderlyingLattice(op);
+			}
+
+			private:
+			mlir::SymbolTableCollection symbolTable;
+			mlir::rlc::FlatFunctionOp function;
+			DataFlowConfig config;
+			std::unique_ptr<DataFlowSolver> solver;
+			ConstraintsAnalysis* analysis;
+		};
+
+	}	 // namespace
+
+
+#define GEN_PASS_DEF_CONSTRAINTSANALYSISPASS
+#include "rlc/dialect/Passes.inc"
+
+	struct ConstraintsAnalysisPass
+			: impl::ConstraintsAnalysisPassBase<ConstraintsAnalysisPass>
+	{
+		using impl::ConstraintsAnalysisPassBase<
+				ConstraintsAnalysisPass>::ConstraintsAnalysisPassBase;
+
+		void runOnOperation() override
+		{
+			//Perform the analysis only if the function returns a boolean
+			for (auto fun : getOperation().getOps<mlir::rlc::FlatFunctionOp>())
+			{
+				
+				if(not mlir::isa<mlir::rlc::BoolType>(fun.getResult().getType().getResult(0)) )
+					continue;
+
+				if(fun.getBlocks().size()<=2)
```

Comment:
why can't we inspect functions with 1 block? 
say
```
f(Int a):
 %bb1:
    ret a
```

Reply:
good question, actually I do not remember why I put that but I think its because all the cases I tested have at least these two blocks in the IR ( for example):

```
^bb0(%arg0: !rlc.int<64>):
    %4 = rlc.ref @rl_can_foo__int64_t_r_bool : (!rlc.int<64>) -> !rlc.bool
    %5 = rlc.call %4 : (!rlc.int<64>) -> !rlc.bool(%arg0) {is_member_call = false} : (!rlc.int<64>) -> !rlc.bool
    rlc.crb %5 !rlc.bool ^bb2 ^bb1
  ^bb1:  // pred: ^bb0
    rlc.abort
```

So I think this check is kinda useless now that I think about it.